README.md
# mongo-cache-example

Example express app that implements mongo-backed cache server with plain HTTP API

## Getting started

### Requirements

- `>=node-8.0.0`
- `>=mongodb-3.4.0`

### Install dependencies

```bash
npm i
```

### Update configuration

In order to update configuration (if needed) you may create `local.js` in `config/` directory with needed variables overrides.

Example:

```js
const config = {};
module.exports = config;

config.cache = {
  size: 100,
  ttl: 3 * 60 * 1000,
};

config.db = {
  url: 'mongodb://example.com:27017',
};
```

### Run server

```bash
npm start
```

## Releasing new version

Use `npm version` command to release new version. It will do all necessary checks and will push it to VCS


CHANGELOG.md
#### 1.0.0 (2018-06-15)

##### Chores

* **package.json:**  revert version ([7e595d42](https://github.com/SkeLLLa/mongo-cache-example/commit/7e595d42caf0c9b66c591c2767856d3ec48befe9))
* **deps:**  add deps ([5af02fc5](https://github.com/SkeLLLa/mongo-cache-example/commit/5af02fc59d9e1a3e44312eda745c6aa2fed57ef7))
* **config:**  add tsconfig ([23c92fc4](https://github.com/SkeLLLa/mongo-cache-example/commit/23c92fc4e390839c425b9cabe3cc5e9edeada328))
* **ignore:**  add ignore files ([edb16ede](https://github.com/SkeLLLa/mongo-cache-example/commit/edb16eded4cea4e5c45619e79412518b9e1a156c))
* **initial:**  initial commit ([9b8131a8](https://github.com/SkeLLLa/mongo-cache-example/commit/9b8131a8dfd7fecaaed5981a6673b80a1a14c963))

##### Documentation Changes

* **readme:**  update readme with getting started info ([983ed2a8](https://github.com/SkeLLLa/mongo-cache-example/commit/983ed2a82cb19d4572513184133753498d7d77cd))
* **changelog:**  add changelog file ([df335fba](https://github.com/SkeLLLa/mongo-cache-example/commit/df335fbaf5d595cd3fa0da04f7209e96b5b36107))
* **task:**  add task pdf ([e9f9881f](https://github.com/SkeLLLa/mongo-cache-example/commit/e9f9881f3cca1b54e3bf17f97d404af276bcdc50))

##### New Features

* **server:**  add server ([e34b3d9d](https://github.com/SkeLLLa/mongo-cache-example/commit/e34b3d9dff89fbe07a6ca5c41a0c00d7c5626b21))
* **db:**  add db module ([1aceac38](https://github.com/SkeLLLa/mongo-cache-example/commit/1aceac3831f2d6c0af904c6bd2dcd1bfcc38a09d))
* **api:**
  *  add api implementation ([7f65b642](https://github.com/SkeLLLa/mongo-cache-example/commit/7f65b642b90790db30dd2008c2400c2b866821fa))
  *  add api docs ([b6096838](https://github.com/SkeLLLa/mongo-cache-example/commit/b60968385d3dc79523306458a32af8dc8808e915))
* **data:**  add random data generator ([7ecfa34c](https://github.com/SkeLLLa/mongo-cache-example/commit/7ecfa34ceb3999d09f8276446628ddee2a620581))
* **config:**  add default config ([fdb8a2bf](https://github.com/SkeLLLa/mongo-cache-example/commit/fdb8a2bfb250242cb3361ead3bdc426f3902c68f))

##### Bug Fixes

* **package.json:**  fix typo in scripts ([8b128df8](https://github.com/SkeLLLa/mongo-cache-example/commit/8b128df83cfa6032d48ccd6121ce2d5fdd953ed2))

##### Refactors

* **changelog:**  revert changelog ([4fe7673c](https://github.com/SkeLLLa/mongo-cache-example/commit/4fe7673c513f5b28f6c252c6d13a6b6da89cef76))
* **package.json:**  update repo name ([c3a6cb4e](https://github.com/SkeLLLa/mongo-cache-example/commit/c3a6cb4e9ebbfb0adacb2b0eb452bde397166030))
* **server:**  add api information ([1a859ef9](https://github.com/SkeLLLa/mongo-cache-example/commit/1a859ef9524de8404d8873e11c01600fdefa7b29))

##### Code Style Changes

* **codestyle:**  add codestyle config ([54868aae](https://github.com/SkeLLLa/mongo-cache-example/commit/54868aaee707d22f8f6dc1439c3282894336cbfc))



lib/db.js
/// <reference types="node" />
/// <reference types="mongodb" />
/// <reference types="config" />

const {MongoClient} = require('mongodb');
const config = require('config').get('db');
let db;
let collection;

const mongo = {
  get db() {
    return db;
  },
  get collection() {
    return collection;
  },
  async connect() {
    try {
      const client = await MongoClient.connect(config.url);
      db = client.db(config.db);
      collection = db.collection(config.collection);
      await collection.createIndex(
        {expireAt: 1},
        {expireAfterSeconds: 0}
      );
      await collection.createIndex('key', {unique: true, dropDups: true});
    } catch (ex) {
      console.error(ex);
      console.log(`Can't connect to DB`);
      process.exit(1);
    }
  },
  close() {
    if (db) {
      db.close();
    }
  },
};

module.exports = mongo;


lib/server.js
/// <reference types="node" />
/// <reference types="mongodb" />
/// <reference types="config" />
/// <reference types="express" />

const express = require('express');
const ui = require('swagger-ui-express');
const apiDoc = require('./api-doc');
const bodyParser = require('body-parser');
const router = require('swagger-express-router');
const api = require('./api');
const config = require('config');
const app = express();
const {connect} = require('./db');

const apiConfig = config.get('api');
const serverConfig = config.get('server');

app.use(
  apiConfig.docsPath,
  ui.serve,
  ui.setup(apiDoc, {
    explorer: true,
  })
);
app.use(bodyParser.text());
router.setUpRoutes(api, app, apiDoc, true);

exports.start = async () => {
  await connect();
  app.listen(serverConfig);

  console.info(
    `Server started. Browse api: http://${serverConfig.host}:${
      serverConfig.port
    }/api-docs`
  );
};


lib/api-doc.js
/// <reference types="node" />
/// <reference types="config" />

const {version, description, name} = require('../package.json');
const config = require('config');
const apiConfig = config.get('api');
const serverConfig = config.get('server');

const api = {
  swagger: '2.0',
  info: {
    title: name,
    description,
    version,
  },
  produces: ['text/plain', 'application/json'],
  host: `${serverConfig.host}:${serverConfig.port}`,
  basePath: apiConfig.basePath,
  paths: {
    '/keys': {
      get: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'getKeys',
        'tags': ['cache'],
        'description': 'Get all keys',
        'produces': 'application/json',
        'parameters': [],
        'responses': {
          200: {
            description: 'All cache keys list',
            schema: {
              type: 'array',
              items: {
                type: 'string',
              },
            },
          },
        },
      },
    },
    '/cache': {
      get: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'getAll',
        'tags': ['cache'],
        'description': 'Get all cache entries',
        'produces': 'application/json',
        'parameters': [],
        'responses': {
          200: {
            description: 'All cache entries list',
            schema: {
              type: 'array',
              items: {
                type: 'object',
                properties: {
                  key: {
                    type: 'string',
                  },
                  value: {
                    type: 'string',
                  },
                  ttl: {
                    type: 'number',
                  },
                },
              },
            },
          },
        },
      },
      delete: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'deleteAll',
        'tags': ['cache'],
        'description': 'Delete all cache entries',
        'parameters': [],
        'responses': {
          204: {
            description: 'Cache cleared',
          },
        },
      },
    },
    '/cache/{key}': {
      get: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'get',
        'tags': ['cache'],
        'description': 'Get cache entry',
        'parameters': [
          {
            name: 'key',
            in: 'path',
            type: 'string',
            required: true,
            description: 'Cache key',
          },
        ],
        'responses': {
          200: {
            description: 'Cache HIT',
            schema: {
              type: 'string',
            },
          },
          201: {
            description: 'Cache MISS. New value created',
            schema: {
              type: 'string',
            },
          },
          400: {
            description: 'Bad request',
            schema: {
              type: 'string',
            },
          },
        },
      },
      put: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'set',
        'tags': ['cache'],
        'description': 'Set cache entry',
        'consumes': ['text/plain'],
        'parameters': [
          {
            name: 'key',
            in: 'path',
            type: 'string',
            required: true,
            description: 'Cache key',
          },
          {
            name: 'ttl',
            in: 'query',
            type: 'number',
            required: false,
            description: 'Cache item ttl',
          },
          {
            name: 'value',
            in: 'body',
            type: 'string',
            consumes: ['text/plain'],
            required: true,
            description: 'Cache value',
          },
        ],
        'responses': {
          201: {
            description: 'Cache item created',
            schema: {
              type: 'string',
            },
          },
          400: {
            description: 'Bad request',
            schema: {
              type: 'string',
            },
          },
        },
      },
      delete: {
        'x-swagger-router-controller': 'cache',
        'operationId': 'delete',
        'tags': ['cache'],
        'description': 'Remove entry from cache',
        'parameters': [
          {
            name: 'key',
            in: 'path',
            type: 'string',
            required: true,
            description: 'Cache key',
          },
        ],
        'responses': {
          204: {
            description: 'Cache item removed',
          },
          400: {
            description: 'Bad request',
            schema: {
              type: 'string',
            },
          },
          404: {
            description: 'Key not found',
            schema: {
              type: 'string',
            },
          },
        },
      },
    },
  },
};

module.exports = api;


lib/api/cache.js
/// <reference types="node" />
/// <reference types="mongodb" />
/// <reference types="express" />

const mongo = require('../db');
const config = require('config').get('cache');
const {getRandomString} = require('../data-generator');

const getValue = (value, ttl = config.ttl) => {
  const now = new Date();
  return {
    updatedAt: now,
    expireAt: new Date(now.getTime() + ttl),
    ttl,
    value,
  };
};

const updateOne = async (key, obj) => {
  const data = await mongo.collection.updateOne(
    {key},
    {$set: obj},
    {
      upsert: true,
    }
  );
  const count = await mongo.collection.count();
  // check if cache count exceeds cache size
  // and delete oldest entry
  if (count > config.size) {
    await deleteOledst();
  }

  return data;
};

const getOne = async (key) => {
  return mongo.collection.findOne({key});
};

const deleteOne = async (key) => {
  return mongo.collection.deleteOne({key});
};

const deleteOledst = async () => {
  const oldest = await mongo.collection.findOne({}, {sort: {updatedAt: 1}});
  await deleteOne(oldest.key);
  return;
};

const deleteAll = async () => {
  return mongo.collection.deleteMany({});
};

const getAll = async (fields = []) => {
  const projection = {_id: 0};
  fields.forEach((it) => {
    projection[it] = 1;
  });
  return mongo.collection.find({}, {projection}).toArray();
};

exports.get = async (req, res) => {
  try {
    const {key} = req.params;
    const {ttl} = req.query;
    if (typeof key === 'undefined') {
      res.status(400).send('Bad request');
      return;
    }
    const data = await getOne(key);
    if (data !== null) {
      console.info(`Cache HIT. "${data.value}"`);
      res.status(200).send(data.value);
      await updateOne(key, getValue(data.value, data.ttl));
      return data.value;
    }
    const newVal = await getRandomString();
    console.info(`Cache MISS. New value created "${newVal}"`);
    await updateOne(key, getValue(newVal, ttl));
    res.status(201).send(newVal);
    return newVal;
  } catch (ex) {
    console.error(ex);
    res.status(500).send('Something went wrong!');
  }
};

exports.getKeys = async (req, res) => {
  try {
    const data = await getAll(['key']);
    const result = data.map((it) => it.key);
    res.status(200).send(result);
    return result;
  } catch (ex) {
    res.status(500).send('Something went wrong!');
  }
};

exports.getAll = async (req, res) => {
  try {
    const data = await getAll(['key', 'value', 'ttl']);
    const result = data.map((it) => {
      return {
        key: it.key,
        value: it.value,
        ttl: it.ttl,
      };
    });
    res.status(200).send(result);
    return result;
  } catch (ex) {
    res.status(500).send('Something went wrong!');
  }
};

exports.set = async (req, res) => {
  try {
    const {key} = req.params;
    const value = req.body;
    const {ttl} = req.query;
    if (typeof key === 'undefined' || typeof value === 'undefined') {
      res.status(400).send('Bad request');
      return;
    }
    await updateOne(key, getValue(value, ttl));
    res.status(201).send('Ok');
  } catch (ex) {
    res.status(500).send('Something went wrong!');
  }
};

exports.delete = async (req, res) => {
  try {
    const {key} = req.params;
    if (typeof key === 'undefined') {
      res.status(400).send('Bad request');
      return;
    }
    const {deletedCount} = await deleteOne(key);
    if (deletedCount) {
      res.status(204).end();
      return;
    }
    res.status(404).end(`Key "${key}" not found`);
  } catch (ex) {
    res.status(500).send('Something went wrong!');
  }
};

exports.deleteAll = async (req, res) => {
  try {
    await deleteAll();
    res.status(204).end();
  } catch (ex) {
    res.status(500).send('Something went wrong!');
  }
};
